spec/ameba/rule/lint/unused_instance_variable_access_spec.cr

```diff
@@ -0,0 +1,64 @@
+require "../../../spec_helper"
+
+module Ameba::Rule::Lint
+  describe UnusedInstanceVariableAccess do
+    subject = UnusedInstanceVariableAccess.new
+
+    it "passes if instance variables are used for assignment" do
+      expect_no_issues subject, <<-CRYSTAL
+        class MyClass
+          foo = @ivar
+        end
+        CRYSTAL
+    end
+
+    it "passes if @type is unused within a macro expression" do
+      expect_no_issues subject, <<-CRYSTAL
+        def foo
+          {% @type %}
+          :bar
+        end
+        CRYSTAL
+    end
+
+    it "passes if an instance variable is used as a target in multi-assignment" do
+      expect_no_issues subject, <<-CRYSTAL
+        class MyClass
+          @foo, @bar = 1, 2
+        end
+        CRYSTAL
+    end
+
+    it "fails if instance variables are unused in void context of class" do
+      expect_issue subject, <<-CRYSTAL
+        class Actor
+          @name : String = "George"
+
+          @name
+        # ^^^^^ error: Value from instance variable access is unused
+        end
+        CRYSTAL
+    end
+
+    it "fails if instance variables are unused in void context of method" do
+      expect_issue subject, <<-'CRYSTAL'
+        def hello : String
+          @name
+        # ^^^^^ error: Value from instance variable access is unused
+
+          "Hello, #{@name}!"
+        end
+        CRYSTAL
+    end
+
+    it "fails if @def is unused within a macro expression" do
+      expect_issue subject, <<-CRYSTAL
+        def foo
+          {% @def %}
+           # ^^^^ error: Value from instance variable access is unused
+          :bar
+        end
+        CRYSTAL
+    end
```

Comment:
```suggestion
    it "fails if instance variable is unused within a macro expression" do
      expect_issue subject, <<-CRYSTAL
        def foo
          {% @bar %}
           # ^^^^ error: Value from instance variable access is unused
          :baz
        end
        CRYSTAL
    end
```

also, please group the related test cases together

Reply:
Actually, after a second thought I think that you can't really use instance variables within macros.

Reply:
You can use these ones https://github.com/crystal-lang/crystal/blob/9b7fd8dc19288f6281814a9bae4b81e75aac32f8/src/compiler/crystal/macros/interpreter.cr#L541